File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/notification/PlaybackNotificationBroadcaster.java
package com.lasthopesoftware.bluewater.client.playback.service.notification;

import com.lasthopesoftware.bluewater.client.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.playback.service.notification.building.BuildNowPlayingNotificationContent;
import com.lasthopesoftware.resources.notifications.control.ControlNotifications;
import com.namehillsoftware.handoff.promises.response.VoidResponse;


public class PlaybackNotificationBroadcaster implements NotifyOfPlaybackEvents {

	private final ControlNotifications notificationsController;
	private final PlaybackNotificationsConfiguration playbackNotificationsConfiguration;
	private final BuildNowPlayingNotificationContent nowPlayingNotificationContentBuilder;

	private final Object notificationSync = new Object();
	private boolean isPlaying;
	private boolean isNotificationStarted;
	private ServiceFile serviceFile;

	public PlaybackNotificationBroadcaster(ControlNotifications notificationsController, PlaybackNotificationsConfiguration playbackNotificationsConfiguration, BuildNowPlayingNotificationContent nowPlayingNotificationContentBuilder) {
		this.notificationsController = notificationsController;
		this.playbackNotificationsConfiguration = playbackNotificationsConfiguration;
		this.nowPlayingNotificationContentBuilder = nowPlayingNotificationContentBuilder;
	}

	@Override
	public void notifyPlaying() {
		isPlaying = true;

		if (serviceFile != null)
			updateNowPlaying(serviceFile);
	}

	@Override
	public void notifyPaused() {
		if (serviceFile == null) {
			notificationsController.stopForegroundNotification(playbackNotificationsConfiguration.getNotificationId());
			return;
		}

		nowPlayingNotificationContentBuilder.promiseNowPlayingNotification(serviceFile, isPlaying = false)
			.then(builder -> {
				notificationsController.notifyBackground(builder.build(), playbackNotificationsConfiguration.getNotificationId());
				return null;
			});
	}

	@Override
	public void notifyStopped() {
		synchronized (notificationSync) {
			isPlaying = false;
			isNotificationStarted = false;
			notificationsController.removeForegroundNotification(playbackNotificationsConfiguration.getNotificationId());
		}
	}

	@Override
	public void notifyPlayingFileChanged(ServiceFile serviceFile) {
		updateNowPlaying(serviceFile);
	}

	private void updateNowPlaying(ServiceFile serviceFile) {
		synchronized (notificationSync) {
			this.serviceFile = serviceFile;

			if (!isNotificationStarted && !isPlaying) return;

			nowPlayingNotificationContentBuilder.promiseNowPlayingNotification(serviceFile, isPlaying)
				.then(new VoidResponse<>(builder -> {
					synchronized (notificationSync) {
						if (!isPlaying) {
							if (!isNotificationStarted) return;

							notificationsController.notifyBackground(builder.build(), playbackNotificationsConfiguration.getNotificationId());
							return;
						}

						isNotificationStarted = true;
						notificationsController.notifyForeground(builder.build(), playbackNotificationsConfiguration.getNotificationId());
					}
				}));
		}
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/service/notification/PlaybackNotificationBroadcaster.kt
package com.lasthopesoftware.bluewater.client.playback.service.notification

import android.app.Notification
import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile
import com.lasthopesoftware.bluewater.client.playback.service.notification.building.BuildNowPlayingNotificationContent
import com.lasthopesoftware.bluewater.client.playback.service.notification.building.BuildPlaybackStartingNotification
import com.lasthopesoftware.bluewater.shared.android.notifications.control.ControlNotifications

class PlaybackNotificationBroadcaster(
	private val notificationsController: ControlNotifications,
	notificationsConfiguration: NotificationsConfiguration,
	private val nowPlayingNotificationContentBuilder: BuildNowPlayingNotificationContent,
	private val playbackStartingNotification: BuildPlaybackStartingNotification
) : NotifyOfPlaybackEvents {

	private val notificationId = notificationsConfiguration.notificationId
	private val notificationSync = Any()

	private var isPlaying = false
	private var isNotificationStarted = false
	private var serviceFile: ServiceFile? = null

	override fun notifyPlaying() {
		isPlaying = true

		serviceFile?.also {
			updateNowPlaying(it)
			return
		}

		playbackStartingNotification.promisePreparedPlaybackStartingNotification()
			.then { builder ->
				synchronized(notificationSync) {
					if (isNotificationStarted) return@then
					isNotificationStarted = true
					notificationsController.notifyForeground(builder.build(), notificationId)
				}
			}
	}

	override fun notifyPaused() {
		if (serviceFile == null) {
			notificationsController.stopForegroundNotification(notificationId)
			return
		}

		nowPlayingNotificationContentBuilder
			.promiseNowPlayingNotification(serviceFile, false.also { isPlaying = it })
			.then { builder ->
				notificationsController.notifyBackground(builder.build(), notificationId)
			}
	}

	override fun notifyInterrupted() {
		if (serviceFile == null) {
			notificationsController.stopForegroundNotification(notificationId)
			return
		}

		nowPlayingNotificationContentBuilder
			.promiseNowPlayingNotification(serviceFile, false.also { isPlaying = it })
			.then { builder ->
				notificationsController.notifyForeground(builder.build(), notificationId)
			}
	}

	override fun notifyStopped() {
		synchronized(notificationSync) {
			isPlaying = false
			isNotificationStarted = false
			notificationsController.removeNotification(notificationId)
		}
	}

	override fun notifyPlayingFileChanged(serviceFile: ServiceFile) {
		updateNowPlaying(serviceFile)
	}

	private fun updateNowPlaying(serviceFile: ServiceFile) {
		fun notify(notification: Notification) {
			when {
				isPlaying -> {
					isNotificationStarted = true
					notificationsController.notifyForeground(notification, notificationId)
				}
				isNotificationStarted -> {
					notificationsController.notifyEither(notification, notificationId)
				}
			}
		}

		synchronized(notificationSync) {
			this.serviceFile = serviceFile

			if (!isNotificationStarted && !isPlaying) return

			val loadingBuilderNotification =
				nowPlayingNotificationContentBuilder.getLoadingNotification(isPlaying).build()

			notify(loadingBuilderNotification)

			nowPlayingNotificationContentBuilder.promiseNowPlayingNotification(serviceFile, isPlaying)
				.then { builder ->
					synchronized(notificationSync) { notify(builder.build()) }
				}
		}
	}
}
